File: src/markers.ts
function nonemptyLines(text: string): string[] {
  return text
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
}

// ONE convention for every marker parser: scan the whole output and take the
// LAST line that matches the marker. Prompts ask for the marker as the final
// line, but models add preamble and trailing prose; anchoring to a fixed
// position (first line, exact last line) silently mis-parsed real outputs —
// including the reconcile escalation valve, which failed OPEN (proceed) on any
// preamble. Last-match wins because a stage's genuine verdict supersedes any
// earlier draft or quoted marker.
function lastMarkerMatch(text: string, pattern: RegExp): RegExpExecArray | null {
  let last: RegExpExecArray | null = null
  for (const line of nonemptyLines(text)) {
    const match = pattern.exec(line)
    if (match) {
      last = match
    }
  }
  return last
}

export type TriageResult = {
  trivial: boolean
  userFacing: boolean
  complexityMarker: boolean
  userFacingMarker: boolean
  // Raw IMPLEMENTER marker value (only requested when an implementer pool is
  // configured). Pure extraction — DEFAULT and unknown names pass through;
  // the conductor resolves against the pool.
  implementer: string | null
}
export type ShipResult = { ok: true } | { ok: false; reason: string }
type DeliveryConfidence = 'low' | 'medium' | 'high' | null
type ParsedDeliveryMeta = {
  confidence: DeliveryConfidence
  reason: string | null
}
export type DeliverySelection = {
  delivery:
    | ({ mode: 'none'; source: 'selected' | 'fallback' } & ParsedDeliveryMeta)
    | ({ mode: 'skill'; skill: string; source: 'selected' } & ParsedDeliveryMeta)
    | ({ mode: 'policy'; policy: string; source: 'selected' } & ParsedDeliveryMeta)
}
export type ReconcileDecision = 'PROCEED' | 'ASK' | null
// The verify-gate doctor's classification of a failure (see remediatePrompt):
// ENV-FIXED — environment problem the doctor repaired (re-run verify);
// ENV-BLOCKED — environment problem it couldn't fix (back off);
// CODE — a real code/test defect (route to the code-fix loop);
// FLAKE — a transient/external failure (back off and retry);
// GATE-MISCONFIGURED — the verify command itself is broken; `gateFix` carries
// the doctor's corrected command (required for the verdict to be usable).
export type RemedyVerdict =
  | 'ENV-FIXED'
  | 'ENV-BLOCKED'
  | 'CODE'
  | 'FLAKE'
  | 'GATE-MISCONFIGURED'
  | null

export function parseShipUrl(text: string): string | null {
  const value = lastMarkerMatch(text, /^URL:\s*(\S+)\s*$/i)?.[1] ?? null
  return value && /^https?:\/\//.test(value) ? value : null
}

export function parseMoot(text: string): boolean | null {
  const value = lastMarkerMatch(text, /^MOOT:\s*(YES|NO)\s*$/i)?.[1]?.toUpperCase()
  return value === 'YES' ? true : value === 'NO' ? false : null
}

export function parseGateFix(text: string): string | null {
  return lastMarkerMatch(text, /^GATE-FIX:\s*(.+)$/i)?.[1]?.trim() ?? null
}

export function parseTriage(text: string): TriageResult {
  const complexity = lastMarkerMatch(
    text,
    /^COMPLEXITY:\s*(TRIVIAL|COMPLEX)\s*$/i
  )?.[1]?.toUpperCase()
  const userFacing = lastMarkerMatch(text, /^USER-FACING:\s*(YES|NO)\s*$/i)?.[1]?.toUpperCase()
  const implementer = lastMarkerMatch(text, /^IMPLEMENTER:\s*(.+)$/i)?.[1]?.trim() ?? null

  return {
    trivial: complexity === 'TRIVIAL',
    userFacing: userFacing === 'YES',
    // Missing markers still resolve (complex / not-user-facing) but callers can
    // now see that the resolution was a default, not a model decision — a missing
    // USER-FACING silently disabled the UX review gate before this.
    complexityMarker: complexity !== undefined,
    userFacingMarker: userFacing !== undefined,
    implementer,
  }
}

export function parseReviewVerdict(text: string): 'PASS' | 'FAIL' | null {
  const value = lastMarkerMatch(text, /^VERDICT:\s*(PASS|FAIL)\s*$/i)?.[1]?.toUpperCase()
  return value === 'PASS' || value === 'FAIL' ? value : null
}

export type ConvergenceVerdict =
  | 'CONTINUE_CODE_FIX'
  | 'RETRY_LATER'
  | 'ASK_HUMAN'
  | 'TERMINAL'
  | null

export function parseConvergenceVerdict(text: string): ConvergenceVerdict {
  const value = lastMarkerMatch(
    text,
    /^VERDICT:\s*(CONTINUE_CODE_FIX|RETRY_LATER|ASK_HUMAN|TERMINAL)\s*$/i
  )?.[1]?.toUpperCase()
  if (
    value === 'CONTINUE_CODE_FIX' ||
    value === 'RETRY_LATER' ||
    value === 'ASK_HUMAN' ||
    value === 'TERMINAL'
  ) {
    return value
  }
  return null
}

export function parseRemedy(text: string): RemedyVerdict {
  const value = lastMarkerMatch(
    text,
    /^VERDICT:\s*(ENV-FIXED|ENV-BLOCKED|CODE|FLAKE|GATE-MISCONFIGURED)\s*$/i
  )?.[1]?.toUpperCase()
  if (
    value === 'ENV-FIXED' ||
    value === 'ENV-BLOCKED' ||
    value === 'CODE' ||
    value === 'FLAKE' ||
    value === 'GATE-MISCONFIGURED'
  ) {
    return value
  }
  return null
}

export function parseShip(text: string): ShipResult {
  const match = lastMarkerMatch(text, /^SHIP:\s*(OK|FAILED)\s*(.*)$/i)
  if (match?.[1]?.toUpperCase() === 'OK') {
    return { ok: true }
  }
  return { ok: false, reason: match?.[2]?.trim() || 'ship did not report success' }
}

function markerValue(text: string, name: string): string | null {
  const match = new RegExp(`^${name}:\\s*(.+)$`, 'im').exec(text)
  return match?.[1]?.trim() ?? null
}

export function parseDeliverySelection(text: string, availableSkills: string[]): DeliverySelection {
  const confidenceText = markerValue(text, 'CONFIDENCE')?.toLowerCase()
  const confidence =
    confidenceText === 'low' || confidenceText === 'medium' || confidenceText === 'high'
      ? confidenceText
      : null
  const reason = markerValue(text, 'REASON')
  const raw = markerValue(text, 'DELIVERY')
  if (!raw) {
    return {
      delivery: {
        mode: 'none',
        source: 'fallback',
        confidence: 'low',
        reason: 'delivery selector did not report a valid DELIVERY marker',
      },
    }
  }

  if (/^none$/i.test(raw)) {
    return {
      delivery: { mode: 'none', source: 'selected', confidence, reason },
    }
  }

  const skill = /^skill\s+([a-z0-9._-]+)$/i.exec(raw)?.[1]
  const canonicalSkill = skill
    ? availableSkills.find((available) => available.toLowerCase() === skill.toLowerCase())
    : null
  if (canonicalSkill) {
    return {
      delivery: {
        mode: 'skill',
        skill: canonicalSkill,
        source: 'selected',
        confidence,
        reason,
      },
    }
  }

  const policy = /^policy\s+(.+)$/i.exec(raw)?.[1]?.trim()
  if (policy) {
    return {
      delivery: { mode: 'policy', policy, source: 'selected', confidence, reason },
    }
  }

  return {
    delivery: {
      mode: 'none',
      source: 'fallback',
      confidence: 'low',
      reason: `delivery selector returned unsupported delivery: ${raw}`,
    },
  }
}

export function parseReconcileDecision(text: string): ReconcileDecision {
  const value = lastMarkerMatch(text, /^DECISION:\s*(PROCEED|ASK)\s*$/i)?.[1]?.toUpperCase()
  if (value === 'PROCEED' || value === 'ASK') {
    return value
  }
  return null
}

File: src/markers.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TriageResult {
    pub trivial: bool,
    pub user_facing: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ShipResult {
    Ok,
    Failed { reason: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ReconcileDecision {
    Proceed,
    Ask,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RemedyVerdict {
    EnvFixed,
    EnvBlocked,
    Code,
    Flake,
}

fn nonempty_lines(text: &str) -> Vec<String> {
    text.lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(ToOwned::to_owned)
        .collect()
}

fn first_nonempty_line(text: &str) -> String {
    nonempty_lines(text).into_iter().next().unwrap_or_default()
}

fn final_nonempty_line(text: &str) -> String {
    nonempty_lines(text).into_iter().last().unwrap_or_default()
}

fn marker_value(
    line: &str,
    marker: &str,
    allowed: &'static [&'static str],
) -> Option<&'static str> {
    let (head, value) = line.split_once(':')?;
    if !head.eq_ignore_ascii_case(marker) {
        return None;
    }
    let value = value.trim();
    allowed
        .iter()
        .find(|candidate| value.eq_ignore_ascii_case(candidate))
        .copied()
}

pub fn parse_triage(text: &str) -> TriageResult {
    let lines = nonempty_lines(text);
    let complexity_line = lines
        .len()
        .checked_sub(2)
        .and_then(|idx| lines.get(idx))
        .map(String::as_str)
        .unwrap_or_default();
    let user_facing_line = lines.last().map(String::as_str).unwrap_or_default();

    TriageResult {
        trivial: marker_value(complexity_line, "COMPLEXITY", &["TRIVIAL", "COMPLEX"])
            == Some("TRIVIAL"),
        user_facing: marker_value(user_facing_line, "USER-FACING", &["YES", "NO"]) == Some("YES"),
    }
}

pub fn parse_review_verdict(text: &str) -> Option<&'static str> {
    match marker_value(&final_nonempty_line(text), "VERDICT", &["PASS", "FAIL"]) {
        Some("PASS") => Some("PASS"),
        Some("FAIL") => Some("FAIL"),
        _ => None,
    }
}

pub fn parse_convergence_verdict(text: &str) -> Option<&'static str> {
    match marker_value(
        &final_nonempty_line(text),
        "VERDICT",
        &["CONTINUE", "STUCK"],
    ) {
        Some("CONTINUE") => Some("CONTINUE"),
        Some("STUCK") => Some("STUCK"),
        _ => None,
    }
}

pub fn parse_remedy(text: &str) -> Option<RemedyVerdict> {
    match marker_value(
        &final_nonempty_line(text),
        "VERDICT",
        &["ENV-FIXED", "ENV-BLOCKED", "CODE", "FLAKE"],
    ) {
        Some("ENV-FIXED") => Some(RemedyVerdict::EnvFixed),
        Some("ENV-BLOCKED") => Some(RemedyVerdict::EnvBlocked),
        Some("CODE") => Some(RemedyVerdict::Code),
        Some("FLAKE") => Some(RemedyVerdict::Flake),
        _ => None,
    }
}

pub fn parse_ship(text: &str) -> ShipResult {
    let line = final_nonempty_line(text);
    if marker_value(&line, "SHIP", &["OK"]) == Some("OK") {
        return ShipResult::Ok;
    }
    let Some((head, value)) = line.split_once(':') else {
        return ShipResult::Failed {
            reason: "ship did not report success".to_string(),
        };
    };
    if !head.eq_ignore_ascii_case("SHIP") {
        return ShipResult::Failed {
            reason: "ship did not report success".to_string(),
        };
    }
    let value = value.trim();
    let upper = value.to_uppercase();
    if !upper.starts_with("FAILED") {
        return ShipResult::Failed {
            reason: "ship did not report success".to_string(),
        };
    }
    let reason = &value["FAILED".len()..];
    let reason = reason.trim();
    ShipResult::Failed {
        reason: if reason.is_empty() {
            "ship did not report success".to_string()
        } else {
            reason.to_string()
        },
    }
}

pub fn parse_reconcile_decision(text: &str) -> Option<ReconcileDecision> {
    match marker_value(&first_nonempty_line(text), "DECISION", &["PROCEED", "ASK"]) {
        Some("PROCEED") => Some(ReconcileDecision::Proceed),
        Some("ASK") => Some(ReconcileDecision::Ask),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_triage_from_final_two_marker_lines() {
        assert_eq!(
            parse_triage(
                "Reasoning is allowed before the markers.\nCOMPLEXITY: TRIVIAL\nUSER-FACING: YES"
            ),
            TriageResult {
                trivial: true,
                user_facing: true,
            }
        );
    }

    #[test]
    fn malformed_triage_defaults_to_complex_not_user_facing() {
        assert_eq!(
            parse_triage("COMPLEXITY: MAYBE\nUSER-FACING: MAYBE"),
            TriageResult {
                trivial: false,
                user_facing: false,
            }
        );
    }

    #[test]
    fn review_verdict_must_be_final() {
        assert_eq!(
            parse_review_verdict("VERDICT: PASS\n\n## Findings\nnone"),
            None
        );
        assert_eq!(
            parse_review_verdict("## Findings\nnone\n\nVERDICT: PASS"),
            Some("PASS")
        );
        assert_eq!(parse_review_verdict("> VERDICT: PASS"), None);
    }

    #[test]
    fn convergence_verdict_must_be_final() {
        assert_eq!(
            parse_convergence_verdict("VERDICT: CONTINUE\n\nsame root cause"),
            None
        );
        assert_eq!(
            parse_convergence_verdict("same root cause\n\nVERDICT: STUCK"),
            Some("STUCK")
        );
        assert_eq!(parse_convergence_verdict("`VERDICT: CONTINUE`"), None);
    }

    #[test]
    fn parses_ship_only_from_final_line() {
        assert_eq!(parse_ship("opened PR\nSHIP: OK"), ShipResult::Ok);
        assert_eq!(
            parse_ship("SHIP: OK\nfollow-up text"),
            ShipResult::Failed {
                reason: "ship did not report success".to_string()
            }
        );
        assert_eq!(
            parse_ship("could not push\nSHIP: FAILED remote rejected"),
            ShipResult::Failed {
                reason: "remote rejected".to_string()
            }
        );
    }

    #[test]
    fn parses_remedy_only_from_final_line() {
        assert_eq!(
            parse_remedy("ran bun install\nSUMMARY: deps were missing\nVERDICT: ENV-FIXED"),
            Some(RemedyVerdict::EnvFixed)
        );
        assert_eq!(
            parse_remedy("SUMMARY: assertion failed\nVERDICT: CODE"),
            Some(RemedyVerdict::Code)
        );
        assert_eq!(parse_remedy("VERDICT: ENV-FIXED\ntrailing note"), None);
        assert_eq!(parse_remedy("VERDICT: MAYBE"), None);
    }

    #[test]
    fn parses_reconcile_only_from_first_nonempty_line() {
        assert_eq!(
            parse_reconcile_decision("DECISION: ASK\n\nWhat should the label be?"),
            Some(ReconcileDecision::Ask)
        );
        assert_eq!(
            parse_reconcile_decision("\nDECISION: PROCEED\nAssuming the default."),
            Some(ReconcileDecision::Proceed)
        );
        assert_eq!(
            parse_reconcile_decision("I might ask.\nDECISION: ASK"),
            None
        );
        assert_eq!(parse_reconcile_decision("> DECISION: ASK"), None);
    }
}
